docs: add section on nodes and actors (#3975)
AztecProtocol/aztec-packages: yellow-paper/docs/decentralization/actors.md

---
title: Nodes and Actors
---

> To analyze the suitability of different node types for various actor types in your system, let's break down the node types and actor types, considering the characteristics, preferences, abilities, and potential motivations of each actor. This will help determine which actors are likely to run which nodes and how they might interact with the Private Execution Environment (PXE).

Reviewer: ```suggestion
To analyze the suitability of different node types for various actor types in your system, let's break down the node types and actor types, considering the characteristics, preferences, abilities, and potential motivations of each actor. This will help determine which actors are likely to run which nodes and how they might interact with the Private eXecution Environment (PXE).
```

😝

## Node Types

The below node types are ordered in resource requirements. The types exists both in the context of Aztec but also other blockchains. All the nodes, participate in the peer-to-peer (p2p) network, but with varying capacity.

1. **Light Node**: 
    - Follows the blockchain but only interested in headers (stores headers).
    - Minimal resource requirements.
    - Can be used to anchor state membership proofs with hard-to-manipulate state root.
2. **Pruned Full Node**:
   - Stores complete state, but only subset of logs and history.
   - Requires less storage than a Full Node.
   - Can be used to simulate transactions and view current state.
2. **Full Node**:

>     - Stores complete state, logs and history.

Reviewer 2: Is the distinction between full and archive nodes:

- History is the contents of blocks and logs stored.
- Historical state is the ability to say leaf index was 'x' at block 'y' and provide corresponding hashpaths.

    - Requires large storage and a stable internet connection.
    - Can be used to simulate transactions and view current state and provide history for new nodes to sync.
3. **Archive Node**:

>     - Stores complete state, logs and history as well as historical state.

Reviewer: By history here you mean "full content" of transactions and not just its state effects?

Author: History mainly being that you have all of the blocks that have been inserted (or you can at least recreate it from what you got).

#### Special nodes
Special nodes are nodes that have specific tasks beyond synchronizing the chain. Namely they participate in block production. In Ethereum they are called validators, but for the sake of briefness, we skip those here, and mainly focus on the Aztec specific nodes.

These nodes are the sequencer and prover nodes. To function they must be connected to a Pruned Full Node, Full Node or Archive Node. 

1. **Sequencer**: 
   - Requires computational power to simulate blocks.
   - Needs a robust network to coordinate proof generation and propagate blocks.
2. **Provers**: 
   - Have high-performance CPUs/GPUs and a lot of memory.
   - Assists in proving zero-knowledge (zk) proofs for block validation.

### Aztec Data Retrieval Capabilities:

> An Aztec node needs to be connected to an Ethereum L1 node to learn about the state of the system. Initially synching can be done this way, but due to the validity proofs, it will also support to download the state from a third party and then match it against the L1 state transitioner to validate that the state is correct. 

Reviewer: This is quite cool. Downloading the full compressed state from something like bittorrent would probably be much faster then asking nodes for individual blobs.

Author: Ye. I guess it is not actually because of the validity proofs (unless you make an aggregated like mina that you are having there as well). But more than you can just match the state roots against what is on L1 😎

As the Aztec node can be connected to different types of underlying network nodes, its ability to retrieve data will vary. Below we outline briefly what the capabilities are for each type of node:
1. **Ethereum Light Node**:

>    - Relies on third party for getting L2 headers and block data.

Reviewer 3: how do they verify?

Author: That is what they are using the light node for, light node have the header you check that the tx is actually in there and then you go from there 
![image](https://github.com/AztecProtocol/aztec-packages/assets/16536249/d26bcb90-82e7-40f0-a498-89777094326e)

### Actor Types

1. **Retail**:
   - Least knowledgeable and least interested in technical details.
   - Prefers simplicity and ease of use.
   - Likely to avoid running any node software or only running if bundled
2. **Enthusiast**: 
   - More knowledgeable and interested in technology than retail.
   - Willing to invest time and resources but not at a professional level.
3. **Power Users**: 
   - Technically proficient and deeply engaged.
   - Likely to have the resources and motivation to run more demanding nodes.
4. **Developers**: 
   - Highly knowledgeable with technical skills.
   - Interested in detailed state and historical data for development and testing.
5. **Self-Validators**: 
   - Aim to participate actively in keeping the blockchain available.
   - Interested in relying on their own nodes and resources (Archive Nodes).
   - Likely to run nodes that contribute directly to the blockchain's operation, like Sequencers or Provers.
6. **Institutions**: 
   - Have significant resources (and potentially technical) capabilities.
   - Interested in robust participation, potentially running Sequencer or Provers for financial gain.

:::info State-Snapshots

> State snapshots is likely made and published by actors in the **Self-validator** or **Institution** category. These are actors that have the resources to run Archive Nodes and are interested in the historical data.

Reviewer: ```suggestion
State snapshots are likely made and published by actors in the **Self-validator** or **Institution** category. These are actors that have the resources to run Archive Nodes and are interested in the historical data.
```

This is placed a bit out-of-context here. Would expand on why we need state snapshots.

Author: State snapshot here refers to the publication of state at a specific point, e.g., you don't have the current, so it is just a snapshot of the state.  Related to your earlier comment.

:::

> ### Private Execution Environment (PXE)

Reviewer: ```suggestion
### Private eXecution Environment (PXE)
```

😝 😝

The Aztec PXE is required to do anything meaningful for an end-user. It is the interface between the user and the Aztec node. The PXE is a private enclave that runs on the end-user's machine.

It is responsible for execution contracts and building proofs for the user's transactions privately. This requires the PXE to be connected to a node that can provide the user's state.

:::info Note-discovery
Note, that as the node will not store secrets for the user. The user relies on some mechanism to figure out "what to lookup", this is what is sometimes referred to as note-discovery: discovering what notes belong to the user.
:::

> As the PXE learns of the users notes, it is stored in the PXE's storage for later usage. Beware that the PXE need to revalidate the membership paths of the notes when they are to be spent which will connect to the node.

Reviewer: ```suggestion
As the PXE learns of the users' notes, it stores them in a database for later usage. Beware that the PXE needs to revalidate the membership paths of the notes when they are to be spent. These paths will be obtained from an Aztec node.
```

### Users

From the end-users perspective, they will be interacting with the PXE, which is then talking to the aztec node and the underlying chain.

```mermaid
graph LR
   A[User] --> B[PXE]
   B --> C[Aztec Node]
   C --> D[Ethereum Node]
```

The type of node the PXE is connected to depends on the actor type. 

:::info Bundling

> For a mobile application to work nicely for the users the PXE will most likely be bundled with a light node on device. This will allow the user to interact with the system without wider knowledge. Settings might allow you to connect the PXE directly to an endpoint for a node for advanced users.

Reviewer: ```suggestion
For a mobile application to work smoothly and for it to not leak information about which note hashes in the note hash tree pertain to notes owned by the users the PXE will most likely be bundled with a light node on device. This will allow the user to interact with the system without wider knowledge. In the settings we might allow you to connect the PXE directly to an external node (this might be relevant for advanced users).
```

> For a mobile application to work nicely for the users the PXE will most likely be bundled with a light node on device. This will allow the user to interact with the system without wider knowledge. Settings might allow you to connect the PXE directly to an endpoint for a node for advanced users.

Reviewer 2: I don't think we would bundle a light node on a user's device. At the top you mention that all nodes would connect to the p2p network. We wouldn't have user's devices do this. The PXE on their device would need to call a remote node.

Author: Light client would only need to get the header, which for our case could use the P2P similar to what happens for most other blockchains or you could abuse L1 as your p2p layer for getting the header. The light client is there to provide you with stronger guarantees around the header than blind trust.

Bundling it is a way of getting users into using more secure structures. And allows us to have the logic for getting a hold of your notes at the node. In the case that you connect to your own full node, you can simply get it to return values from its local db based on whatever it checked with tags etc. And if light-node the lightnode can call out to services that might exists and validate the returned values against its headers for anchoring. That way the PXE don't need to deal with figuring out how to get the data beyond "just ask the node", and node can then implement what it desires. Similarly to what you can do for things like eth, the PXE or wallet don't care what type of node it is connected to, just that the node provides the expected answers.

Reviewer 4: I think Phil's original point is valid, though, right? I think there are two categories here: a user who only runs a PXE and has to query everything about the blockchain from a 3rd-party server; and a user who runs a PXE & light node on their device. I do think both types of user will exist.

Reviewer: @iAmMichaelConnor the desirable end-game of blockchains is that everyone runs a validating light clients and I feel like given that we only need to fetch the headers and that we will have a much lower block frequency then Ethereum then it actually might be feasible to aim for having users use a validating light client (of Aztec) always, no?

Reviewer 4: It would be nice if everyone ran a node. But pragmatically, I don't think that'll happen, so recongnising this class of user seems sensible to me.
If I'm using aztec on my phone, do I want to drain my battery running a light node, or do I want to run the bare minimum software, and make requests to a node only when I need to? I might even be running a full node at home, which I can make requests to. But a resource constrained device doesn't want to stay in sync with the blockchain, I don't think?

Reviewer 2: I suppose I'm not entirely sure what we envisage a 'light' node doing. In my mind, the minimum  amount of functionality for something to be called a 'node' is:

1. Providing world state (hash paths etc).
2. Proving block headers.
3. Providing access to the tx pool's p2p network for transaction submission.

You would never run the above on a mobile device. The p2p chatter and hashing would sap your battery and consume all of your bandwidth.

If what's required is merely a way to independently verify block headers so that I know the node I am connecting to isn't lying to me can that be achieved by configuring the PXE (which is on a user's mobile) to connect to a trusted L1 node and query that information directly?

Reviewer: In general light nodes are nodes which have the necessary information to verify that they are not being lied to but don't store any other info. Both on Ethereum and in our case it would be just fetching and hashing the headers. By hashing the headers and hashing the archive tree users get the archive root which they then can verify against the Rollup contract on L1.

Author: Updated document to hopefully answer this more ahead of time.

:::

1. **Mainstream User**: 

>    - Running the bare minimum. PXE bundled with a light node.

Reviewer 2: Same as above. I don't think these people would be running a node connected to the p2p network.

   - Likely using low-power devices like mobile phones or old laptops.

>    - Relies heavily on third-party services (Infura for L1 and "OMR" for L2 state)

Reviewer: ```suggestion
   - Relies heavily on third-party services (Infura for L1 and some form of private information retrieval service (e.g. Oblivious Message Retrieval service) for L2 state)
```

2. **Enthusiast**:
   - While not fully retail, they are still not interested in running a full node. Essentially the same as retail from end-user perspective.
3. **Power Users**:

>    - Will likely run a (pruned) full node and connect the PXE to that.

Reviewer: ```suggestion
   - Will likely run a (pruned) full node, with their PXE connected to it.
```

   - Relies on third-party services for historical L1 data, but not for L2 data.
4. **Developers**:

>    - Similar to power user, but more likely to run non-pruned node and connect PXE to that.

Reviewer: ```suggestion
   - Similar to power user, but more likely to run non-pruned node with their PXE connected to it.
```

   - Non-pruned used for other data access
5. **Self-Validators**:

>    - Likely to run a full node or archive node and connect PXE to that.

Reviewer: ```suggestion
   - Likely to run a full node or archive node with their PXE connected to it..
```

6. **Institutions**:

>    - Likely to use a managed service for running multiple nodes for better P2P connectivity for improved propagation.

Reviewer: ```suggestion
   - Likely to use a managed service for running multiple nodes, for better P2P connectivity, and as a result, improved data propagation.
```

>    - Connecting the PXE to this managed service.

Reviewer: ```suggestion
   - Connecting their PXE to this managed service.
```

---
title: Nodes and Actors
---

To analyze the suitability of different node types for various actor types in the system, let's break down the node types and actor types, considering the characteristics, preferences, abilities, and potential motivations of each actor. This will help determine which actors are likely to run which nodes and how they might interact with the Private eXecution Environment (PXE).

## Node Types

The below node types are ordered in resource requirements. The types exists both in the context of Aztec but also other blockchains. All the nodes, participate in the peer-to-peer (p2p) network, but with varying capacity.

1. **Light Node**: 

>     - Follows the blockchain but only interested in headers (stores headers).

Reviewer 4: Since we have 3 blockchains (the Aztec L2 p2p network, the Ethereum p2p network, and the DA p2p network), we should specify which of those p2p networks each of these participants needs to track. Do they need to run a node for all 3 networks? If not, which do they need to run, etc.

Reviewer 4: Edit, I see a section below on Ethereum nodes. Which kinds of ethereum nodes does a user need to run, if they want to run a particular type of aztec node? Or maybe it's easier to say what type of ethereum node is _not_ sufficient, if you want to run a particular type of aztec node? This info could perhaps form part of the table.

>     - Minimal resource requirements.

Reviewer 4: Perhaps we need to link to Zac's doc in the ../cryptography/ section, which discusses target resource requirements.

Reviewer 4: Whoops, I forgot to "start a review". Sorry for the spam.

    - Can be used to anchor state membership proofs with hard-to-manipulate state root.
2. **Pruned Full Node**:
   - Stores complete state, but only subset of logs and history.
   - Requires less storage than a Full Node.
   - Can be used to simulate transactions and view current state.
2. **Full Node**:
    - Stores complete state, logs and history.
    - Requires large storage and a stable internet connection.
    - Can be used to simulate transactions and view current state and provide history for new nodes to sync.
3. **Archive Node**:

>     - Stores complete state, logs and history as well as historical state. 

Reviewer 4: What's the difference between "history" and "historical state"?

An Aztec node needs to be connected to an Ethereum L1 node to learn about the state of the system. Initially synching can be done this way, but since the L1 can always be used as the "trust-anchor" for the state, an Aztec node can retrieve state from a third party and then validate it against the values of L1.

As the Aztec node can be connected to different types of underlying network nodes, its ability to retrieve data will vary. Below we outline briefly what the capabilities are for each type of node:
1. **Ethereum Light Node**:
   - Relies on third party for getting L2 headers and block data.
2. **Ethereum Pruned Full Node**:
   - Can lookup recent L2 headers and blocks from own data.
   - Relies on third party for getting historical blocks
3. **Ethereum Full Node**:
   - Can lookup L2 headers from own data.
   - Can lookup recent L2 blocks from own data
   - Relies on third party for getting historical blocks
4. **Ethereum Archive Node**:
   - Can lookup L2 headers and blocks from own data.

### Actor Types

1. **Mainstream User**:
   - Least knowledgeable and least interested in technical details.
   - Prefers simplicity and ease of use.

>    - Likely to avoid running any node software or only running if bundled

Reviewer 4: I'd maybe break this further into users who are willing to make public queries to a 3rd-party node; and users who insist on making privacy-preserving queries to special 3rd-party nodes.

2. **Enthusiast**: 
   - More knowledgeable and interested in technology than the mainstream user.
   - Willing to invest time and resources but not at a professional level.
3. **Power Users**: 
   - Technically proficient and deeply engaged.
   - Likely to have the resources and motivation to run more demanding nodes.
4. **Developers**: 
   - Highly knowledgeable with technical skills.
   - Interested in detailed state and historical data for development and testing.
5. **Self-Validators**: 
   - Aim to participate actively in keeping the blockchain available.

>    - Interested in relying on their own nodes and resources (Archive Nodes).

Reviewer 4: A self validator doesn't need to run an archive node, surely?

:::info Note-discovery
Note: As the node will not store secrets for the user, the user relies on some mechanism to figure out "what to lookup". This is what is sometimes referred to as note-discovery: discovering what notes belong to the user.
:::

As the PXE learns of the user's notes, it stores them in a database for later usage. Beware that the PXE needs to revalidate the membership paths of the notes when they are to be spent. These paths will be obtained from an Aztec node.

### Users

From the end-users perspective, they will be interacting with the PXE, which is then talking to the aztec node and the underlying chain.

```mermaid
graph LR
   A[User] --> B[PXE]
   B --> C[Aztec Node]
   C --> D[Ethereum Node]
```

The type of node the PXE is connected to depends on the actor type. 

:::info Bundling
For a mobile application to work nicely for the users the PXE will most likely be bundled with a light node on device. This will allow the user to interact with the system without wider knowledge. Settings might allow you to connect the PXE directly to an endpoint for a node for advanced users.
:::

> 1. **Mainstream User**: 

Reviewer 4: We have two lists of user types (with the same user types, but with different info in each list). Might it be possible to combine these two lists into one list?

   - Stores only the headers.
   - Querying any state not in the header is done by requesting the data from a third party, e.g. Infura or other nodes in the p2p network. Responses are validated with the headers as a trust anchor.
   - Storage requirements typically measured in MB's (< 1GB).
   - Synchronization time is typically measured in minutes.
2. **Full Node**:
   - Receive and validate blocks (header and body) from the p2p network.
   - Stores the complete active state of the chain
   - Typically stores recent state history (last couple of hours is common)
   - Typically stores all blocks since genesis (some pruning might be done)
   - Can respond to queries of current and recent state
   - Storage requirements typically measured in GB's (< 1TB)
   - Synchronization time is typically measured in hours/days.
3. **Archive Node**:
   - Receive and validate blocks (header and body) from the p2p network
   - Stores the full state history of the chain
   - Stores all blocks since genesis
   - Can respond to queries of state at any point in time

>    - Storage requirements typically measured in TB's

Author: Changed this, and the MB's and GB's as well.

   - Synchronization time is typically measured in hours/days.

> Beyond these node types, there are also nodes that participate in the block production and rely on full or archive nodes to extend the chain. In Ethereum these are called validators, but really, any of the nodes above can be used to validate the headers and full/archive the blocks as well. The Ethereum validator is really just their block producer. 

Author: Should be easier now.

   - Currently don't care about technicalities, just want to use the system.
   - Will likely not run any type of node, unless it is bundled with their wallet so they don't even know it is there.
   - Generally don't care about trusting Infura or other third parties blindly.
2. **Enthusiast**: 
   - More knowledgeable and interested in technology than the mainstream user.
   - Willing to invest time and resources but not at a professional level.
   - Likely to run a light node for the extra security.
3. **Power Users**: 
   - Technically proficient and deeply engaged.
   - Likely to have the resources and motivation to run more demanding nodes, and potentially participate in block production.
4. **Developers**: 
   - Highly knowledgeable with technical skills.
   - Interested in detailed state and historical data for development and testing.
5. **Idealists**: 
   - Want maximal autonomy, and are willing to invest resources to achieve it.
   - Will rely entirely on their own nodes and resources for serving their queries, current or historical by running an Archive Node.
   - Likely to run nodes that contribute directly to the blockchain's operation as part of block production.

>    - Possibly willing to store all blobs

Reviewer 5: ```suggestion
   - Possibly willing to store all blobs.
```

6. **Institutions**: 
   - Have significant resources (and potentially technical) capabilities.
   - Interested in robust participation, and potentially participate in block production for financial gain.

## Aztec Relevant

In the following section, we are following the assumption that Aztec is running on top of Ethereum and using some DA to publish its data.

Beyond the [state](./../state/index.md) of the chain an Aztec node also maintain a database of encrypted data and their tags - an index of tags. The index is used when a user wish to retrieve notes that belong to them. The index is build by the node as it receive new blocks from the network.

> While the full index will allow the node to serve all users, it could be configured to only build the index for a subset of the notes. For example, a personal node could be configured to only build the index for the notes that belong to the owner based on their [tag derivation](./../private-message-delivery/note-discovery.md). This would allow the node to serve the owner, but not other users.

Reviewer 5: How would the node serve other users? Wouldn't it need to be able to decrypt notes sent to them?

Author: I extended the document a bit here

- Aztec nodes that are not running dependency-minimized can rely on third parties for body retrieval and use Ethereum as a "trust-anchor". Here meaning that the Aztec node could retrieve the full state from IPFS and then validate it against the headers it receives from the Ethereum node - allowing a quick synchronization process. Since the index is not directly part of state, its validity cannot be validated as simple as the state. The index will have to be built from the blocks (validating individual decrypted messages can be done against the state).

An Aztec light-node is a bit different, since it does not store the actual state *nor* index, only the headers. This means that it will follow the same model of Ethereum light-nodes, where it can retrieve data by asking third parties for it and then validating that it is in the state (anchored by the headers stored). 

:::info Aztec-specific Ultra Light nodes

> For users following the validating bridge as the canonical rollup, the bridge IS the light node, so they can get away with not storing anything, and just download the header data from the bridge. Since the [archive](./../state/archive.md) is stored in the bridge, an Aztec ultra-light node can validate membership proofs against it without needing the exact header (given sufficient membership proofs are provided). This essentially enable instant sync for ultra-light nodes -> if you either run an Ethereum node or trust a third party.

Author: Ye it is pretty cool. You don't even need to look for the event of the new block, you can just query the archive root directly, so similar to a `balanceOf` lookup that you run once in a while.

:::

### Responding to Queries

In the following, we will briefly outline how the different type of Aztec node will respond to different kinds of queries. Namely we will consider the following queries:
- Requesting current public state and membership proofs
- Requesting historical public state and membership proofs
- Requesting current membership proof of note hash, nullifier, l1 to l2 message or contract 
- Requesting historical membership proof of note hash, nullifier, l1 to l2 message or contract 
- Requesting encrypted notes for a given tag

#### Light-Node

As mentioned, we will be retrieving data from a third party, and then validating it against the headers at our disposal. Note that if we have the latest archive hash, a response can include an inclusion in a header and in the archive, thereby allowing us to validate the response against the archive as well.

> If we don't care about other people seeing what data we are retrieving, we can simply ask the third parties directly. However, when dealing with private information (e.g. notes), you need to be careful about how you retrieve it - you need to use some form of private information retrieval (PIR) service. 

Reviewer 5: Here they wouldn't be retrieving the notes directly but rather the merkle path required to produce an inclusion proof, right?

Author: Extended the document a bit here as well.

While it is responsible for executing contracts and building proofs for the user's transactions privately, it needs data from the node to do so. It can be connected to any of the above types of nodes, inheriting their assumptions. From the point of view of the PXE, it is connected to a node, and it does not really care which one.

When requesting encrypted notes from the node the PXE will decrypt it and store the decrypted notes in its own database. This database is used to build proofs for the user's transactions. 

From most end users points of view, the PXE will be very close to a wallet.

### Bundling the PXE with an Aztec ultra-light node

A deployment of the PXE could be bundled together with an aztec ultra-light node which in turn is connected to "some" Ethereum node. The ethereum node could simply be infura, and then the ultra-light node handles its connections to the OMR service separately. 

This setup can be made very light-weight and easy to use for the end-user, who might not be aware that they are running a node at all.

> The reasoning that the ultra-light node should be bundled with an internal ultra-light node as default instead of simple an RPC endpoint is to avoid users "missing" the OMR services and leak private information.

Reviewer: ```suggestion
The reasoning that the ultra-light node should be bundled with an internal ultra-light node by default instead of simply using an RPC endpoint is to avoid users "missing" the OMR services and leak private information.
```

---
title: Nodes and Actors
---

To analyze the suitability of different node types for various actor types in the system, let's break down the node types and actor types, considering the characteristics, preferences, abilities, and potential motivations of each actor. 
This will help determine which actors are likely to run which nodes and how they might interact with the Private eXecution Environment (PXE).

## Background

Before diving into Aztec specific data, we take a look at general blockchain nodes. 
It should help us have a shared base, which will later be useful for comparing the Aztec setups, where multiple nodes are involved due to its construction.

### Node Types

The below node types are ordered in resource requirements, and exists for most blockchains. 
All the nodes participate in the peer-to-peer (p2p) network but with varying capacity.

> 1. **Light Node**:

Reviewer 2: I'm still unsure of the purpose of a 'light' node. With regards to the p2p network, I wasn't planning on broadcasting the block headers, nodes will just read them from L1 which is the source of truth. The p2p network only needs to propagate transactions.

Author: Light nodes are to not blindly trust something like infura. 
The section is not worrying around the specific type of blockchain, but general. 

But since the l1 contract we have is a validating light node, you can use it as you p2p connection.
